compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -65,43 +65,89 @@ class Semantic {
   /** An object with unknown initialization status */
   case object Cold extends Value
 
+  sealed abstract class Addr extends Value {
+    def klass: ClassSymbol
+  }
+
   /** A reference to the object under initialization pointed by `this`
    */
-  case class ThisRef(klass: ClassSymbol) extends Value
+  case class ThisRef(klass: ClassSymbol) extends Addr
 
   /** An object with all fields initialized but reaches objects under initialization
    *
    *  We need to restrict nesting levels of `outer` to finitize the domain.
    */
-  case class Warm(klass: ClassSymbol, outer: Value) extends Value
+  case class Warm(klass: ClassSymbol, outer: Value) extends Addr
 
   /** A function value */
-  case class Fun(expr: Tree, thisV: ThisRef | Warm, klass: ClassSymbol) extends Value
+  case class Fun(expr: Tree, thisV: Addr, klass: ClassSymbol) extends Value
 
   /** A value which represents a set of addresses
    *
    * It comes from `if` expressions.
    */
-  case class RefSet(refs: List[Warm | Fun | ThisRef]) extends Value
+  case class RefSet(refs: List[Fun | Addr]) extends Value
 
   // end of value definition
 
-  /** The current object under initialization
+  /** The abstract object which stores value about its fields and immediate outers.
+   *
+   *  Semantically it suffices to store the outer for `klass`. We cache other outers
+   *  for performance reasons.
    *
    *  Note: Object is NOT a value.
    */
-  case class Objekt(klass: ClassSymbol, val fields: mutable.Map[Symbol, Value]) {
-    val promotedValues = mutable.Set.empty[Value]
-  }
+  case class Objekt(klass: ClassSymbol, fields: mutable.Map[Symbol, Value], outers: mutable.Map[ClassSymbol, Value])
 
   /** Abstract heap stores abstract objects
    *
    *  As in the OOPSLA paper, the abstract heap is monotonistic.
    *
    *  This is only one object we need to care about, hence it's just `Objekt`.
    */
-  type Heap = Objekt
-  def heap(using h: Heap): Heap = h
+  object Heap {
```

Comment:
Before this change, the `Heap` contained only one object, the object at whose constructor we begin the analysis with `init`.

After this change, the `Heap` contains more than one object. Which objects are they? How do we decide which concrete objects are represented in the `Heap` and which concrete objects are represented just by abstract values like `Hot` or `Cold`? What is the intended concretization of `Warm(C, outer)` as an `Addr`, which concrete object(s) does it represent? Does a `Heap` always contain exactly one `ThisRef` or can it contain multiple `ThisRef`s with different values of `klass`? If it can contain multiple `ThisRef`s, what is their concretization?

Reply:
Thank you @olhotak. These are good theoretical concerns. Indeed, the refactoring blurs the theoretical structure of the analysis.

I claimed that in essence this PR is the same as master, but I was not clear on why they are the same, and how we arrive at the refactoring. Inspired by your questions above, I think we can see the refactoring in steps, which should answer the questions above:

**1. Introduce Summary for `Warm`**

First, to enable caching of outers for warn objects (and simplify the logic in `resolveThis`), we would like to store the immediate outers of all super-classes in a map, let's call it _summary_:

``` Scala
type Summary = Map[Warm, Objekt]
```

Note that, in the above we use `Objekt` that caches the fields in addition to the outers. The rationale is, why not go a step further to cache fields as well? This way, `Value.select` can avoid dealing with the low-level `cache` for expressions. And in the future, if class parameters can be non-hot, they can easily be handled by storing them as fields.

At this point, we still only have one object in the heap, the _summary_ and _heap_ are two different entities. In essence, it's no different from the master. 

**2. Avoid code duplication**

As you insightfully pointed out in https://github.com/lampepfl/dotty/pull/12495#discussion_r634745509, we should avoid the code duplication in `Value.select` and `Value.call` for `ThisRef` and `Warm`.

As `Warm` is a key to a map, to unify them, we would make `ThisRef` a key to the same map as well:

``` Scala
trait Key extends Value
case class ThisRef(...) extends Key
case class Warm(...) extends Key

type Summary = Map[Key, Objekt]
```

To make sure that the heap only contains on `ThisRef` as key, we remove `ThisRef` from the summary after checking each class. This refactoring would still be the safe as the master.

But why bother removing the key from the summary, if the key is never reachable? So it's fine to keep multiple `ThisRef` keys in the summary -- conceptually we may think the unreachable key is garbage collected.

However, there is a small semantic different between `ThisRef` and `Warm`: selecting a missing field in `ThisRef` is an error, while the field should be computed on-demand for `Warm`. That's the only specialized logic we need.

This overloading usage of summary is directly inspired by your comment in https://github.com/lampepfl/dotty/pull/12495#discussion_r634745509. Overloading of store-like structures happens not only in engineering, but also exists in theoretical work. For example, in [_Abstracting abstract machines_][aam], it uses `Store` to hold both continuations and closures (Section 2.3).

[aam]: https://dl.acm.org/doi/10.1145/1863543.1863553

**3. Renaming**

As the summary contains abstract objects: 

- `ThisRef` refers to objects which may have uninitialized fields
- `Warm` refers to objects whose fields are all initialized but may reach objects with empty fields

Why not rename `key` as `Addr` and `Summary` as `Heap`, which corresponds concrete semantics? Finishing this step leads to the current PR:

``` Scala
trait Addr extends Value
case class ThisRef(...) extends Addr
case class Warm(...) extends Addr

type Heap = Map[Addr, Objekt]
```

This PR also use opaque types for `Heap` instead of type aliases. But such refactoring is cosmetic.

To summarize, the refactoring is inspired by your comments in https://github.com/lampepfl/dotty/pull/12495#discussion_r634745509 and benefited a lot from our previous discussions and the insights from the OOPSLA paper (in OOPSLA we do `this` substitution instead of having cache for `Warm`). The refactoring might have made the theoretical structure less clear, but from the engineering perspective, it seems to be a win.

Reply:
I'll try to explain the fundamental problem. `ThisRef` and `Warm` are abstractions that represent some number of concrete objects. If an `Addr` can represent more than one concrete object, then it is unsound to call `updateField` or `updateOuter` on that `Addr`. I understand that `ThisRef(C)` represents a unique concrete object, so that's fine. But we need to be sure that `Warm(C, outer)` also represents a unique concrete object before we can use it as an `Addr`. This is why I am asking *which* concrete object `Warm(C, outer)` represents.

My understanding of `Warm(C, outer)` is that it represents *all* objects of class `C` whose fields have all been assigned and whose `outer` has the given abstract value. If this understanding is correct, then this change introduces unsoundness, because there can be multiple such concrete objects. If this understanding is not correct, then I would like to correct it by identifying the unique concrete object that `Warm(C, outer)` represents.

Reply:
Would it be correct to understand that `Warm(C, outer)`, when used as an `Addr` for the `Heap`, does not represent *all* warm objects of class `C`, but only *the* unique object on which we invoke `init` to analyze its constructor, in the case when that object is not hot because its outer is not hot?

Reply:
> My understanding of Warm(C, outer) is that it represents all objects of class C whose fields have all been assigned and whose outer has the given abstract value. If this understanding is correct, then this change introduces unsoundness, because there can be multiple such concrete objects. If this understanding is not correct, then I would like to correct it by identifying the unique concrete object that Warm(C, outer) represents.

To be short, it's not unique and I believe it's sound. I don't think being unique is a necessary requirement in all analyses, and definitely not in our case. One example is type analysis, where we abstract the type of a field by the type of the initializer. Then we reject the program if later a value of a different type is assigned to the field.

The same holds for our analysis, the field abstractions cannot be changed by mutation, and it's uniquely determined by the specific value of `Warm`. The `updateField` and `updateOuter` are in essence cache operations --- the same values can be computed with `eval` or fetched from the low-level expression-based cache. All those values are determined by a specific choice of `Warm`. As all objects that are abstracted by the same `Warm` have the same abstract field values, it is safe to compute, cache, and reuse them. 

From anther perspective, this analysis is equivalent to the OOPSLA type-and-effect system (but with better engineering qualities). Any unsoundness example will in turn be a counter-example to the OOPSLA paper.

Reply:
> Would it be correct to understand that `Warm(C, outer)`, when used as an `Addr` for the `Heap`, does not represent _all_ warm objects of class `C`, but only _the_ unique object on which we invoke `init` to analyze its constructor, in the case when that object is not hot because its outer is not hot?

`Warm(C, outer)` is an abstraction for all objects of `C` with the same `outer` at a specific time point during execution of the program. For example, given the following program:

```Scala
class C {
    class B() { ... }
    val a = new B
    val b = new B
    // here
}
```

There are at least two objects above abstracted by `Warm(C, ThisRef(C))` after field `b`. The reason why we can abstract with the same warm value is that they are indistinguishable in the abstract domain --- all field values and outers agree.

Reply:
Based on the discussion above, would it be correct to state the following invariants?

1. The analysis never queries `heap(ThisRef(D))` where `D` is any class other than the class on which `init` was launched. Although the `heap` may contain such `ThisRef`s, they are stale and it would be an error to ask about them.
2. The analysis never queries `heap(ThisRef(C)).outers`. It would be an error to do so.
3. `heap(Warm(C,outer)).field(s) == eval(s.defTree.rhs, Warm(C, outer), C)`
4. `heap(Warm(C,outer)).outers(D) == outerValue(?, Warm(C, outer), D, ?)`

In other words, `heap(ThisRef(C))` maintains the state of the object on which `init` was launched, but `heap(Warm)` is only a cache of `eval` and `outerValue`.

In that case, 3 and 4 are non-trivial results and should have at least some informal proof (not necessarily in the code, of course, but if these are lemmas to be proven externally, they should at least be stated in comments in the code). In particular, are 3 and 4 preserved if `outer` is `Warm` but we then to early promotion of it?

It would also be helpful to list, somewhere, what all the possible values of `outer` in a `Warm` are. Are all `Value`s possible, or only `Warm` and `ThisRef(C)` where `C` is the class on which `init` was launched?

Reply:
> 1. The analysis never queries heap(ThisRef(D)) where D is any class other than the class on which init was launched. Although the heap may contain such ThisRefs, they are stale and it would be an error to ask about them.

Yes, this is correct.

> 2. The analysis never queries heap(ThisRef(C)).outers. It would be an error to do so.

There is flexibility here: either we handle outers for `ThisRef` uniformly as `Warm`, in spite of knowing that they can only   be `Hot`. Or, in `resolveThis`, we have a special case for `ThisRef`. Both works well. Current PR uses the latter. But for uniformity, it's better to do the former.

> 3. heap(Warm(C,outer)).field(s) == eval(s.defTree.rhs, Warm(C, outer), C)

Yes, that's how we compute the value for fields of warm objects -- the place where we call `updateField`. Thus it holds by construction.

> 4. heap(Warm(C,outer)).outers(D) == outerValue(?, Warm(C, outer), D, ?)

Yes, that's how we compute the outers for warm objects  -- the place where we call `updateOuter`. Thus it holds by construction.

> In other words, heap(ThisRef(C)) maintains the state of the object on which init was launched, but heap(Warm) is only a cache of eval and outerValue.

Yes, that's correct. But `heap(ThisRef(C))` can also be regarded as a cache of the abstract values for the fields, so that later access does not need re-computation.

> In particular, are 3 and 4 preserved if outer is Warm but we then to early promotion of it?

3 and 4 hold by construction, thus they are not affected by early promotion. Early promotion does not change summaries for `Warm`.

> It would also be helpful to list, somewhere, what all the possible values of outer in a Warm are. Are all Values possible, or only Warm and ThisRef(C) where C is the class on which init was launched?

That's a good suggestion to improve the documentation of `Warm`. The implicit invariant here is that the `outer` can only be `RefSet`, `Warm` and `Cold`. It should not be `Hot` nor `Fun`.

Reply:
For invariants 3 and 4, the concern is not that the summary (the LHS of the invariant) would change, but that the result of `eval` and `outerValue` (the RHS of the invariant) could change after a change to the fields of the object on which `init` was launched. Specifically, what if the `rhs` of a field of an inner class depends somehow on the outer class, so evaluating the `rhs` changes after the outer class becomes more initialized?

For invariant 3, since `eval` already uses `cache`, why do we need to cache its result again in `heap`?

Reply:
I guess what's confusing me is that the behaviour of `eval` depends on how many fields are in `heap(ThisRef(C)).fields`, but it presumably should not depend on how many fields are in `heap(Warm(C,outer)).fields` or on `heap(???).outers`. Understanding the inputs that `eval` depends on is important to understanding how the analysis works. The confusion is that things that `eval` does and does not depend on are mixed together in `heap`.

Reply:
``` Scala
class C {
    class Base { def m: C }
    class MIsCold extends Base { def m = C.this }
    class MIsHot extends Base { def m = new C }
    class B(base: Base) { val field = base.m }
    val a = new B(new MIsHot)
    val b = new B(new MIsCold)
    // here
    leak(a.field)
    leak(b.field)
    val c = ???
}
```

At `// here`, what is `heap(Warm(B,ThisRef(C))).fields(field)`? For `a` it should be `Hot` but for `b` it should be `Cold`, shouldn't it?

Reply:
Or is it the case that although the type declaration makes `fields` and `outers` mutable for all `Addr` in the heap, we
1. mutate them only during the execution of `init`, and
2. during the execution of `init(thisV = addr)` we mutate only `heap(addr)`, and not the heap of any other addresses.
Then after `init(thisV = addr)` finishes, `heap(addr)` is never mutated anymore. We use it only in a read-only manner to determine what the state was at the end of the constructor of the object represented by `addr`.

Is that right?

Reply:
> For invariants 3 and 4, the concern is not that the summary (the LHS of the invariant) would change, but that the result of `eval` and `outerValue` (the RHS of the invariant) could change after a change to the fields of the object on which `init` was launched. Specifically, what if the `rhs` of a field of an inner class depends somehow on the outer class, so evaluating the `rhs` changes after the outer class becomes more initialized?

This relies on a fundamental idea of the analysis: the changes in the heap don't have any impact on the result of `eval` --- the result of an expression only depends on `thisV` (and `Env` once we have the extension). That's why we can do cache for expressions without having `heap` as part of the key.

That's also the idea in the type-and-effect system where we can create summaries for fields and methods without caring about how many fields are initialized.

> 
> For invariant 3, since `eval` already uses `cache`, why do we need to cache its result again in `heap`?

The reason is a bit hidden in the long comment, the reason is uniformity for engineering benefits, and support future extension to non-hot class parameters:

"Note that, in the above we use Objekt that caches the fields in addition to the outers. The rationale is, why not go a step further to cache fields as well? This way, Value.select can avoid dealing with the low-level cache for expressions. And in the future, if class parameters can be non-hot, they can easily be handled by storing them as fields."

Reply:
> I guess what's confusing me is that the behaviour of eval depends on how many fields are in heap(ThisRef(C)).fields, but it presumably should not depend on how many fields are in heap(Warm(C,outer)).fields or on heap(???).outers. Understanding the inputs that eval depends on is important to understanding how the analysis works. The confusion is that things that eval does and does not depend on are mixed together in heap.

Strictly speaking, the result of `eval` does not depend on the heap for a given valid program, neither `heap(ThisRef(C))` nor `heap(Warm(C,outer))`. Otherwise, the cache has to make the heap part of the key --- which is one of the most important design insights. Similarly, in the type-and-effect system, the summary for an expression does not care about the initialization state of `this`.

Reply:
> ```scala
> class C {
>     class Base { def m: C }
>     class MIsCold extends Base { def m = C.this }
>     class MIsHot extends Base { def m = new C }
>     class B(base: Base) { val field = base.m }
>     val a = new B(new MIsHot)
>     val b = new B(new MIsCold)
>     // here
>     leak(a.field)
>     leak(b.field)
>     val c = ???
> }
> ```
> 
> At `// here`, what is `heap(Warm(B,ThisRef(C))).fields(field)`? For `a` it should be `Hot` but for `b` it should be `Cold`, shouldn't it?

This depends on what abstract domains we are working on. In the current abstract domain, all class parameters must be hot, thus the abstraction for `field` is always `Hot`.

One we extend the domain to allow non-hot class parameters:

```
case class Warm(klass: ClassSymbol, outer: Value, ctor: Symbol, args: List[Value]) extends Addr
```

Then the abstraction for `field` is dependent on the abstract value of `args`. Here we need to widen `args` the same way as we widen `outer` to finitize the addresses to ensure termination, good performance, and usability.

Reply:
> Or is it the case that although the type declaration makes `fields` and `outers` mutable for all `Addr` in the heap, we
> 
> 1. mutate them only during the execution of `init`, and

Yes, that's the only place where `updateOuter` and `updateField` is called. And because of the opaque types, they cannot be accidentally mutated.

> 2. during the execution of `init(thisV = addr)` we mutate only `heap(addr)`, and not the heap of any other addresses.
>    Then after `init(thisV = addr)` finishes, `heap(addr)` is never mutated anymore. We use it only in a read-only manner to determine what the state was at the end of the constructor of the object represented by `addr`.
> 

Yes, that's correct. It's in that sense it's equivalent to the type-and-effect system. In particular, in an assignment, we don't update the heap. This makes it different from typical analysis, where we do a union on the heap for assignment.

Reply:
> This relies on a fundamental idea of the analysis: the changes in the heap don't have any impact on the result of `eval` --- the result of an expression only depends on `thisV` (and `Env` once we have the extension). That's why we can do cache for expressions without having `heap` as part of the key.

I don't understand this point. `eval` calls `Value.select` on values other than `thisV`, and `select` calls `heap(value).fields`.

Reply:
I was not clear, by `thisV` I mean the value for `this`, not `ThisRef`.

Reply:
Note that all the values stored in the heap depend on `thisV`, so the heap is just a cache or summary.